action/container/container.go

```diff
@@ -23,66 +26,126 @@ var (
 // SetupOpts congregates options for Setup function
 // None of the values can be empty string, and mountContainerDir cannot be '.' or '/'
 type SetupOpts struct {
-	ContainerURL      string // URL or name of docker container (name-only will try to look for containers in ghcr.io/9elements/firmware-action
+	ContainerURL      string // URL or name of docker container
 	MountHostDir      string // Directory from host to mount into container
 	MountContainerDir string // Where to mount ^^^ host directory inside container
 	WorkdirContainer  string // Workdir of the container, specified by GITHUB_WORKSPACE environment variable
 }
 
+// Validate the data in struct
+func (opts SetupOpts) Validate() error {
+	pc, _, _, _ := runtime.Caller(0)
+	funcName := logging.TrimNameFunction(pc)
+
+	// None of the directories can be empty string
+	var err error
+	err = nil
+	if opts.MountContainerDir == "" {
+		err = errDirectoryNotSpecified
+		slog.Error(
+			"mountpoint path cannot be empty string",
+			slog.String("function", funcName),
+			slog.String("suggestion", "specify where the host directory should be mounted in the container"),
+			slog.Any("error", err),
+		)
+	}
+	if opts.MountHostDir == "" {
+		err = errDirectoryNotSpecified
+		slog.Error(
+			"host directory path for mounting cannot be empty string",
+			slog.String("function", funcName),
+			slog.String("suggestion", "specify which host directory will be mounted into the container"),
+			slog.Any("error", err),
+		)
+	}
+	if opts.WorkdirContainer == "" {
+		err = errDirectoryNotSpecified
+		slog.Error(
+			"WORKDIR cannot be empty string",
+			slog.String("function", funcName),
+			slog.String("suggestion", "specify working directory for the container"),
+			slog.Any("error", err),
+		)
+	}
+
+	// The mount target directory in container must not be root
+	if opts.MountContainerDir == "." || opts.MountContainerDir == "/" {
+		err = errDirectoryInvalid
+		slog.Error(
+			"container mountpoint cannot be '.' or '/'",
+			slog.String("function", funcName),
+			slog.String("suggestion", "ping another directory, preferably absolute path"),
+			slog.Any("error", err),
+		)
+	}
+	return err
+}
+
 // Setup for setting up a Docker container via dagger
 func Setup(ctx context.Context, client *dagger.Client, opts *SetupOpts, dockerfileDirectoryPath string) (*dagger.Container, error) {
 	// dockerfileDirectoryPath allows to use Dockerfile and build locally,
 	//   which is handy for testing changes to said Dockerfile without the need to
 	//   have the container uploaded into package registry
 
-	// None of the directories can be empty string
-	for _, val := range []string{opts.MountContainerDir, opts.MountHostDir, opts.WorkdirContainer} {
-		if val == "" {
-			return nil, errDirectoryNotSpecified
-		}
-	}
+	pc, _, _, _ := runtime.Caller(0)
+	funcName := logging.TrimNameFunction(pc)
```

Comment:
This feels like it could rather be part of a slog handler as it feels boilerplate-y to have this on top of every function.

Reply:
I was kinda hoping to avoid making handler :/ but the thought of having this on top of every function is less desirable. Valid point